commit 74f4d096b0c30d6db143097dd86773a72dae5580 (refs/changes/53/1453/19)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-08-01T22:03:58+00:00 (2 years, 2 months ago)
    
    MORAY-390 Upgrade Moray to node-fast v2
    MORAY-273 Moray should dump core on uncaught exceptions
    MORAY-335 Moray crashes on bad fast CRC
    MORAY-336 Moray hangs on bad RPC method name
    MORAY-360 Moray throwing TypeErrors, AssertionErrors for some bad input
    MORAY-404 Moray server needs to validate its RPC arguments
    Reviewed by: David Pacheco <dap@joyent.com>
    Reviewed by: Kody Kantor <kody.kantor@joyent.com>
    Reviewed by: Patrick Mooney <patrick.mooney@joyent.com>

diff --git a/Makefile b/Makefile
index cd868ad..0f85d15 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -51,7 +51,7 @@ CLEAN_FILES	+= node_modules $(SHRINKWRAP) cscope.files \
 #
 
 NODE_PREBUILT_TAG	= zone
-NODE_PREBUILT_VERSION	:= v0.10.24
+NODE_PREBUILT_VERSION	:= v0.10.48
 NODE_PREBUILT_IMAGE = fd2cc906-8938-11e3-beab-4359c665ac99
 
 # RELENG-341: no npm cache is making builds unreliable
diff --git a/README.md b/README.md
index c0e7c2c..d7fbc20 100644
--- a/README.md
+++ b/README.md
@@ -179,6 +179,7 @@ Postgres database without using Manatee. A single function is exported,
 - `bindip`, the IP address to bind to
 - `audit`, a boolean indicating whether to log the result and duration of all
   requests
+- `kangPort`, the port the Kang server should listen on
 - `standalone`, an object specifying the standalone server's configuration:
     * `pg`, an object which specifies the Postgres client pool confguration:
         - `queryTimeout`, how long (in milliseconds) before a query is timed out
diff --git a/boot/setup.sh b/boot/setup.sh
index 62eaeac..7e2d88e 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -117,6 +117,7 @@ function setup_moray {
     local ports
     for (( i=1; i<=$moray_instances; i++ )); do
         ports[$i]=`expr 2020 + $i`
+        kangs[$i]=`expr 3020 + $i`
     done
 
     #Regenerate the registrar config with the real ports included
@@ -154,10 +155,13 @@ function setup_moray {
 
     #moray instances
     local moray_xml_in=$SVC_ROOT/smf/manifests/moray.xml.in
-    for port in "${ports[@]}"; do
+    for (( i=1; i<=$moray_instances; i++ )); do
+        local port=${ports[$i]}
+        local kang=${kangs[$i]}
         local moray_instance="moray-$port"
         local moray_xml_out=$SVC_ROOT/smf/manifests/moray-$port.xml
         sed -e "s#@@MORAY_PORT@@#$port#g" \
+            -e "s#@@KANG_PORT@@#$kang#g" \
             -e "s#@@MORAY_INSTANCE_NAME@@#$moray_instance#g" \
             $moray_xml_in  > $moray_xml_out || \
             fatal "could not process $moray_xml_in to $moray_xml_out"
diff --git a/docs/index.md b/docs/index.md
index 7aad8b5..e6eac0f 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -239,7 +239,7 @@ Returns the configuration for all buckets.
      client.listBuckets(function (err, buckets) {
         assert.ifError(err);
         console.log(util.inspect(buckets));
-    });
+     });
 
 ### Inputs
 
diff --git a/etc/config.coal.json b/etc/config.coal.json
index 0b5d727..afd606a 100644
--- a/etc/config.coal.json
+++ b/etc/config.coal.json
@@ -1,6 +1,5 @@
 {
     "port": 2020,
-    "audit": false,
     "bunyan": {
         "level": "info"
     },
diff --git a/etc/config.coal.manta.json b/etc/config.coal.manta.json
index d49c327..10cc8be 100644
--- a/etc/config.coal.manta.json
+++ b/etc/config.coal.manta.json
@@ -7,7 +7,6 @@
     }
   },
   "numWorkers": 0,
-  "audit": false,
   "manatee": {
     "manatee": {
       "path": "/manatee/1.moray.coal.joyent.us/election",
diff --git a/etc/config.lab.json b/etc/config.lab.json
index 360e95b..04d4446 100644
--- a/etc/config.lab.json
+++ b/etc/config.lab.json
@@ -7,7 +7,6 @@
                         "type": "udp"
                 }
         },
-        "audit": false,
         "numWorkers": 0,
         "manatee": {
                 "path": "/manatee/sdc/election",
diff --git a/etc/config.local.json b/etc/config.local.json
index 4fc8ce8..8973f57 100644
--- a/etc/config.local.json
+++ b/etc/config.local.json
@@ -1,6 +1,5 @@
 {
         "port": 2020,
-        "audit": false,
         "bunyan": {
                 "level": "info"
         },
diff --git a/etc/config.standalone.json b/etc/config.standalone.json
index 470de45..c24d9ca 100644
--- a/etc/config.standalone.json
+++ b/etc/config.standalone.json
@@ -4,7 +4,6 @@
     "bunyan": {
         "level": "info"
     },
-    "audit": false,
     "standalone": {
         "pg": {
             "connectTimeout": 3000,
diff --git a/lib/buckets/common.js b/lib/buckets/common.js
index 9898366..442ba40 100644
--- a/lib/buckets/common.js
+++ b/lib/buckets/common.js
@@ -23,6 +23,7 @@ var NotFunctionError = mod_errors.NotFunctionError;
 
 var typeToPg = require('../pg').typeToPg;
 
+var mod_schema = require('../schema');
 
 
 ///--- Globals
@@ -43,7 +44,11 @@ var BUCKET_NAME_RE = /^[a-zA-Z]\w{0,62}$/;
 var INDEX_NAME_RE =
     /^(_[a-zA-Z0-9]+|[a-zA-Z][_a-zA-Z0-9]*[a-zA-Z0-9]|[a-zA-Z])$/;
 
-var RESERVED_BUCKETS = ['moray', 'search'];
+var RESERVED_BUCKETS = [
+    'buckets_config',
+    'moray',
+    'search'
+];
 
 var RESERVED_INDEXES = [
     '_etag',
@@ -133,69 +138,31 @@ function createIndexes(opts, cb) {
 
 
 function validateIndexes(schema) {
-    var i, j, k, k2, keys, msg, sub, subKeys;
+    var i, k, keys;
 
     keys = Object.keys(schema);
     for (i = 0; i < keys.length; i++) {
         k = keys[i];
-        if (typeof (k) !== 'string') {
-            throw new InvalidBucketConfigError('keys must be ' +
-                                               'strings');
-        }
-        if (typeof (schema[k]) !== 'object') {
-            throw new InvalidBucketConfigError('values must be ' +
-                                               'objects');
-        }
 
         if (!INDEX_NAME_RE.test(k)) {
-            throw new InvalidBucketConfigError(
+            return new InvalidBucketConfigError(
                 'invalid index name "' + k + '"');
         }
 
         var kLC = k.toLowerCase();
 
         if (RESERVED_INDEXES.indexOf(kLC) !== -1) {
-            throw new InvalidBucketConfigError(
+            return new InvalidBucketConfigError(
                 'index name "' + k + '" is reserved for use by Moray');
         }
 
         if (/^_*moray/.test(kLC)) {
-            throw new InvalidBucketConfigError(
+            return new InvalidBucketConfigError(
                 'index names cannot start with "moray": "' + k + '"');
         }
-
-        sub = schema[k];
-        subKeys = Object.keys(sub);
-        for (j = 0; j < subKeys.length; j++) {
-            k2 = subKeys[j];
-            switch (k2) {
-            case 'type':
-                if (sub[k2] !== 'string' &&
-                    sub[k2] !== 'number' &&
-                    sub[k2] !== 'boolean' &&
-                    sub[k2] !== 'ip' &&
-                    sub[k2] !== 'subnet' &&
-                    sub[k2] !== '[string]' &&
-                    sub[k2] !== '[number]' &&
-                    sub[k2] !== '[boolean]' &&
-                    sub[k2] !== '[ip]' &&
-                    sub[k2] !== '[subnet]') {
-                    msg = k + '.type is invalid';
-                    throw new InvalidBucketConfigError(msg);
-                }
-                break;
-            case 'unique':
-                if (typeof (sub[k2]) !== 'boolean') {
-                    msg = k + '.unique must be boolean';
-                    throw new InvalidBucketConfigError(msg);
-                }
-                break;
-            default:
-                msg = k + '.' + k2 + ' is invalid';
-                throw new InvalidBucketConfigError(msg);
-            }
-        }
     }
+
+    return null;
 }
 
 
@@ -205,6 +172,11 @@ function validateBucket(req, cb) {
 
     log.debug('validate: entered (%j)', bucket);
 
+    var err = mod_schema.validateBucket(bucket);
+    if (err !== null) {
+        return (cb(err));
+    }
+
     bucket.index = bucket.index || {};
     bucket.post = bucket.post || [];
     bucket.pre = bucket.pre || [];
@@ -231,15 +203,6 @@ function validateBucket(req, cb) {
     if (RESERVED_BUCKETS.indexOf(bucket.name) !== -1)
         return (cb(new InvalidBucketNameError(bucket.name)));
 
-    if (typeof (bucket.index) !== 'object' ||
-        Array.isArray(bucket.index)) {
-        return (cb(new InvalidBucketConfigError('index is not ' +
-                                                'an object')));
-    }
-
-    if (!Array.isArray(bucket.post))
-        return (cb(new NotFunctionError('post')));
-
     try {
         assert.arrayOfFunc(bucket.post);
     } catch (e) {
@@ -247,9 +210,6 @@ function validateBucket(req, cb) {
         return (cb(new NotFunctionError('post')));
     }
 
-    if (!Array.isArray(bucket.pre))
-        return (cb(new NotFunctionError('pre')));
-
     try {
         assert.arrayOfFunc(bucket.pre);
     } catch (e) {
@@ -257,20 +217,9 @@ function validateBucket(req, cb) {
         return (cb(new NotFunctionError('pre')));
     }
 
-    try {
-        validateIndexes(bucket.index);
-    } catch (e) {
-        return (cb(e));
-    }
-
-    if (typeof (bucket.options) !== 'object') {
-        return (cb(new InvalidBucketConfigError('options is not ' +
-                                                'an object')));
-    }
-
-    if (typeof (bucket.options.version) !== 'number') {
-        return (cb(new InvalidBucketConfigError('options.version ' +
-                                                'is not a number')));
+    err = validateIndexes(bucket.index);
+    if (err !== null) {
+        return (cb(err));
     }
 
     log.debug('validate: done');
diff --git a/lib/buckets/creat.js b/lib/buckets/creat.js
index 3537109..9a4c6b6 100644
--- a/lib/buckets/creat.js
+++ b/lib/buckets/creat.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -18,6 +18,25 @@ var BucketConflictError = require('../errors').BucketConflictError;
 var InternalError = require('../errors').InternalError;
 
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'config', type: 'object' },
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    control.getPGHandleAndTransaction,
+    common.validateBucket,
+    insertConfig,
+    createSequence,
+    createLockingSerial,
+    createTable,
+    createIndexes
+];
+
+
 ///--- Handlers
 
 function insertConfig(req, cb) {
@@ -196,10 +215,18 @@ function createIndexes(req, cb) {
 
 function creat(options) {
     control.assertOptions(options);
-    var route = 'createBucket';
 
-    function _creat(name, cfg, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _creat(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var name = argv[0];
+        var cfg = argv[1];
+        var opts = argv[2];
+
+        var req = control.buildReq(opts, rpc, options);
         var bucket = {
             name: name,
             index: cfg.index || {},
@@ -214,20 +241,11 @@ function creat(options) {
             bucket: name,
             cfg: cfg,
             opts: opts
-        }, route + ': entered');
+        }, 'createBucket: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true),
-                common.validateBucket,
-                insertConfig,
-                createSequence,
-                createLockingSerial,
-                createTable,
-                createIndexes
-            ]
+            funcs: PIPELINE
         });
     }
 
diff --git a/lib/buckets/del.js b/lib/buckets/del.js
index 59375e2..44e5c14 100644
--- a/lib/buckets/del.js
+++ b/lib/buckets/del.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -20,6 +20,25 @@ var InvalidBucketNameError = mod_errors.InvalidBucketNameError;
 var BucketNotFoundError = mod_errors.BucketNotFoundError;
 
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    control.getPGHandleAndTransaction,
+    validate,
+    checkExists,
+    deleteConfig,
+    dropTable,
+    dropSequence,
+    dropLockingSerial,
+    common.shootdownBucket
+];
+
+
 ///--- Handlers
 
 function validate(req, cb) {
@@ -28,7 +47,7 @@ function validate(req, cb) {
 
     log.debug('validate: entered (%j)', bucket);
     if (common.RESERVED_BUCKETS.indexOf(bucket.name) !== -1) {
-        cb(new InvalidBucketNameError(bucket));
+        cb(new InvalidBucketNameError(bucket.name));
         return;
     }
 
@@ -170,10 +189,17 @@ function dropLockingSerial(req, cb) {
 
 function del(options) {
     control.assertOptions(options);
-    var route = 'delBucket';
 
-    function _del(bucket, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _del(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var bucket = argv[0];
+        var opts = argv[1];
+
+        var req = control.buildReq(opts, rpc, options);
         req.bucket = {
             name: bucket
         };
@@ -181,21 +207,11 @@ function del(options) {
         req.log.debug({
             bucket: bucket,
             opts: opts
-        }, '%s: entered', route);
+        }, 'delBucket: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true),
-                validate,
-                checkExists,
-                deleteConfig,
-                dropTable,
-                dropSequence,
-                dropLockingSerial,
-                common.shootdownBucket
-            ]
+            funcs: PIPELINE
         });
     }
 
diff --git a/lib/buckets/get.js b/lib/buckets/get.js
index b0ba3dc..76e79e4 100644
--- a/lib/buckets/get.js
+++ b/lib/buckets/get.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -14,6 +14,18 @@ var BucketNotFoundError = require('../errors').BucketNotFoundError;
 var control = require('../control');
 
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'options', type: 'options' },
+    { name: 'bucket', type: 'string' }
+];
+
+var PIPELINE = [
+    control.getPGHandle,
+    loadBucket
+];
+
 
 ///--- Handlers
 
@@ -61,21 +73,29 @@ function loadBucket(req, cb) {
 
 function get(options) {
     control.assertOptions(options);
-    var route = 'getBucket';
 
-    function _get(opts, bucket, res) {
-        var req = control.buildReq(opts, res, options);
+    function _get(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var opts = argv[0];
+        var bucket = argv[1];
+
+        var req = control.buildReq(opts, rpc, options);
         req.bucket = {
             name: bucket
         };
 
+        req.log.debug({
+            bucket: bucket,
+            opts: opts
+        }, 'getBucket: entered');
+
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {read:true}),
-                loadBucket
-            ],
+            funcs: PIPELINE,
             cbOutput: function () { return req.bucket; }
         });
     }
diff --git a/lib/buckets/list.js b/lib/buckets/list.js
index 1e9d8d3..84cdf70 100644
--- a/lib/buckets/list.js
+++ b/lib/buckets/list.js
@@ -5,13 +5,23 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
-require('../errors');
 var control = require('../control');
 
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    control.getPGHandle,
+    loadBuckets
+];
+
 
 ///--- Handlers
 
@@ -34,7 +44,7 @@ function loadBuckets(req, cb) {
     q.on('row', function (r) {
         r.options = r.options || {};
         r.options.version = r.options.version || 0;
-        req.res.write(r);
+        req.rpc.write(r);
     });
 
     q.once('end', function () {
@@ -48,18 +58,24 @@ function loadBuckets(req, cb) {
 
 function list(options) {
     control.assertOptions(options);
-    var route = 'listBuckets';
 
-    function _list(opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _list(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var opts = argv[0];
+
+        var req = control.buildReq(opts, rpc, options);
+
+        req.log.debug({
+            opts: opts
+        }, 'listBuckets: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {read:true}),
-                loadBuckets
-            ]
+            funcs: PIPELINE
         });
     }
 
diff --git a/lib/buckets/update.js b/lib/buckets/update.js
index 25c5985..65ba175 100644
--- a/lib/buckets/update.js
+++ b/lib/buckets/update.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016, Joyent, Inc.
+ * Copyright 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -25,6 +25,29 @@ var BucketVersionError = mod_errors.BucketVersionError;
 var typeToPg = require('../pg').typeToPg;
 
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'name', type: 'string' },
+    { name: 'config', type: 'object' },
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    control.getPGHandleAndTransaction,
+    common.validateBucket,
+    loadBucket,
+    ensureReindexProperty,
+    calculateDiff,
+    ensureRowVer,
+    updateConfig,
+    dropColumns,
+    addColumns,
+    createIndexes,
+    createUniqueIndexes,
+    common.shootdownBucket
+];
+
 
 ///--- Helpers
 
@@ -449,10 +472,18 @@ function createUniqueIndexes(req, cb) {
 
 function update(options) {
     control.assertOptions(options);
-    var route = 'updateBucket';
 
-    function _update(name, cfg, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _update(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var name = argv[0];
+        var cfg = argv[1];
+        var opts = argv[2];
+
+        var req = control.buildReq(opts, rpc, options);
         var bucket = {
             name: name,
             index: cfg.index || {},
@@ -467,25 +498,11 @@ function update(options) {
             bucket: name,
             cfg: cfg,
             opts: opts
-        }, '%s: entered', route);
+        }, 'updateBucket: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true),
-                common.validateBucket,
-                loadBucket,
-                ensureReindexProperty,
-                calculateDiff,
-                ensureRowVer,
-                updateConfig,
-                dropColumns,
-                addColumns,
-                createIndexes,
-                createUniqueIndexes,
-                common.shootdownBucket
-            ]
+            funcs: PIPELINE
         });
     }
 
diff --git a/lib/control.js b/lib/control.js
index 21f5737..316fb38 100644
--- a/lib/control.js
+++ b/lib/control.js
@@ -5,19 +5,33 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 
-
 var assert = require('assert-plus');
 var vasync = require('vasync');
 var libuuid = require('libuuid');
 
-require('./errors');
+var mod_errors = require('./errors');
+var mod_schema = require('./schema');
+
+var InvocationError = mod_errors.InvocationError;
 var pgError = require('./pg').pgError;
 var dtrace = require('./dtrace');
 
+///--- Globals
+
+var SCHEMA_TO_DESCR = {
+    'integer': 'a nonnegative integer',
+    'string': 'a nonempty string',
+    'requests': 'an array of valid request objects',
+    'array': 'an array',
+    'object': 'an object',
+    'options': 'a valid options object'
+};
+
+
 ///--- API
 
 
@@ -28,53 +42,51 @@ function assertOptions(options) {
     assert.object(options.bucketCache, 'options.bucketCache');
 }
 
-function _getPGHandle(route, manatee, params, begin) {
-    params = params || {};
-    var method = manatee.pg.bind(manatee);
-    if (begin) {
-        // run pg.begin before returning control
-        method = manatee.start.bind(manatee);
+
+function _getPGHandleAfter(req, cb) {
+    function done(startErr, pg) {
+        if (startErr) {
+            req.log.debug(startErr, '%s: no DB handle', req.route);
+            cb(startErr);
+        } else {
+            if (req.opts.timeout)
+                pg.setTimeout(req.opts.timeout);
+            req.pg = pg;
+            cb(null);
+        }
     }
 
-    return function getPGHandle(req, cb) {
-        method(params, function (startErr, pg) {
-            if (startErr) {
-                req.log.debug(startErr, '%s: no DB handle', route);
-                cb(startErr);
-            } else {
-                if (req.opts.timeout)
-                    pg.setTimeout(req.opts.timeout);
-                req.pg = pg;
-                cb(null);
-            }
-        });
-    };
+    return done;
 }
 
-function releasePGHandle(req) {
-    if (req.pg) {
-        req.pg.release();
-        req.pg = null;
-    }
+
+function getPGHandle(req, cb) {
+    req.manatee.pg(_getPGHandleAfter(req, cb));
+}
+
+
+function getPGHandleAndTransaction(req, cb) {
+    // run pg.begin before returning control
+    req.manatee.start(_getPGHandleAfter(req, cb));
 }
 
 function handlerPipeline(options) {
     assert.object(options, 'options');
-    assert.string(options.route, 'options.route');
     assert.arrayOfFunc(options.funcs, 'options.funcs');
     assert.object(options.req, 'options.req');
-    assert.object(options.req.res, 'options.req.res');
-    // Modifiers for res.end() and dtrace.fire() output
+    assert.object(options.req.rpc, 'options.req.rpc');
+    assert.string(options.req.route, 'options.req.route');
+    // Modifiers for rpc.end() and dtrace.fire() output
     assert.optionalFunc(options.cbOutput, 'options.cbOutput');
     assert.optionalFunc(options.cbProbe, 'options.cbProbe');
 
-    var route = options.route;
     var req = options.req;
-    var res = req.res;
+    var route = req.route;
+    var rpc = req.rpc;
     var log = req.log;
 
     var cbOutput = options.cbOutput || function () { return null; };
-    var cbProbe = options.cbProbe || function () { return ([res.msgid]); };
+    var cbProbe = options.cbProbe || function () { return ([req.msgid]); };
 
     function pipelineCallback(err) {
         var probe = route.toLowerCase() + '-done';
@@ -94,39 +106,39 @@ function handlerPipeline(options) {
                 req.pg.rollback();
                 req.pg = null;
             }
-            res.end(pgError(err));
+            rpc.fail(pgError(err));
             done();
             return;
         }
+
         req.pg.commit(function (err2) {
             req.pg = null;
             if (err2) {
                 log.debug(err2, '%s: failed', route);
-                res.end(pgError(err2));
+                rpc.fail(pgError(err2));
             } else {
                 var result = cbOutput();
                 if (result) {
                     log.debug({result: result}, '%s: done', route);
-                    res.end(result);
+                    rpc.end(result);
                 } else {
                     log.debug('%s: done', route);
-                    res.end();
+                    rpc.end();
                 }
             }
             done();
         });
-        return;
     }
 
-    vasync.forEachPipeline({
+    req.pipeline = vasync.forEachPipeline({
         inputs: options.funcs,
         func: function (handler, cb) {
             dtrace.fire('handler-start', function () {
-                return [res.msgid, route, handler.name, req.req_id];
+                return [req.msgid, route, handler.name, req.req_id];
             });
             handler(options.req, function (err) {
                 dtrace.fire('handler-done', function () {
-                    return [res.msgid, route, handler.name];
+                    return [req.msgid, route, handler.name];
                 });
                 cb(err);
             });
@@ -134,30 +146,124 @@ function handlerPipeline(options) {
     }, pipelineCallback);
 }
 
-function buildReq(opts, res, serverOpts) {
+function buildReq(opts, rpc, serverOpts) {
     assert.object(opts);
-    assert.object(res);
+    assert.object(rpc);
     assert.object(serverOpts);
 
+    /*
+     * Note that the 'msgid' below is the request ID generated by and for the
+     * Fast protocol, and unrelated to the 'req_id' sent by Moray consumers
+     * to help identify related requests across different Triton/Manta services.
+     *
+     * The 'connId' and 'msgid' aren't used by Moray (except in DTrace probes),
+     * but are attached to the logger and to the req object to help correlate
+     * log messages and objects in crash dumps. The names are chosen to match
+     * the same names used in Fast logs for easier analysis.
+     */
+    var connid = rpc.connectionId();
+    var msgid = rpc.requestId();
+    var route = rpc.methodName();
+    var reqid = opts.req_id || libuuid.create();
+
+    var log = serverOpts.log.child({
+        connId: connid,
+        msgid: msgid,
+        route: route,
+        req_id: reqid
+    });
+
     var req = {
-        req_id: opts.req_id || libuuid.create(),
-        msgid: res.msgid,
+        log: log,
+        req_id: reqid,
+        route: route,
+        connId: connid,
+        msgid: msgid,
         opts: opts,
-        res: res,
+        rpc: rpc,
+        manatee: serverOpts.manatee,
         bucketCache: serverOpts.bucketCache
+
+        /*
+         * A number of RPCs add their own fields to this object. Some common
+         * ones used across a number of RPCs are:
+         *
+         * - "pg", the Manatee client
+         * - "pipeline", the vasync status object for the RPC's operations
+         * - "bucket", the configuration of the affected/searched bucket
+         * - "key", the key being affected by this RPC
+         * - "where", an object representing the WHERE clause for a SQL query,
+         *   containing two fields:
+         *
+         *     - "clause", the actual "WHERE ..." to use in the SQL string
+         *     - "args", the values needed by the clause's $1, $2, etc.
+         *
+         * - "_etag", the new etag for objects affected by the RPC
+         * - "_count", the number of objects touched by an RPC
+         *
+         * Not everything is listed here, since each RPC also has its own unique
+         * set of fields that it stores on this object.
+         */
     };
-    req.log = serverOpts.log.child({
-        req_id: req.req_id
-    });
+
     return req;
 }
 
+
+/*
+ * Validate that the correct number of arguments are provided to an RPC, and are
+ * of the expected type. If the arguments are incorrect, then this function will
+ * handle failing the RPC with an appropriate error, and return 'true'. If the
+ * arguments are okay, then the function will return 'false', and the RPC can
+ * continue normally.
+ *
+ * - rpc: The node-fast rpc object
+ * - argv: The array of arguments provided to the RPC (obtained from rpc.argv())
+ * - types: An array of { name, type } objects describing the arguments expected
+ *   by this RPC, and their types
+ *
+ * Valid types to check for are:
+ *
+ * - array: Check that the argument is an Array
+ * - string: Check that the argument is a nonempty String
+ * - integer: Check that the argument is a nonnegative integer.
+ * - object: Check that the argument is a non-null object.
+ */
+function invalidArgs(rpc, argv, types) {
+    var route = rpc.methodName();
+    var len = types.length;
+
+    if (argv.length !== len) {
+        rpc.fail(new InvocationError(
+            '%s expects %d argument%s', route, len, len === 1 ? '' : 's'));
+        return true;
+    }
+
+    for (var i = 0; i < len; i++) {
+        var name = types[i].name;
+        var type = types[i].type;
+        var val = argv[i];
+
+        var err = mod_schema.validateArgument(name, type, val);
+        if (err !== null) {
+            rpc.fail(new InvocationError(err,
+                '%s expects "%s" (args[%d]) to be %s',
+                route, name, i, SCHEMA_TO_DESCR[type]));
+            return true;
+        }
+    }
+
+    return false;
+}
+
+
 ///--- Exports
 
 module.exports = {
     assertOptions: assertOptions,
-    getPGHandle: _getPGHandle,
-    releasePGHandle: releasePGHandle,
+    getPGHandle: getPGHandle,
+    getPGHandleAndTransaction: getPGHandleAndTransaction,
     handlerPipeline: handlerPipeline,
+    invalidArgs: invalidArgs,
     buildReq: buildReq
 };
diff --git a/lib/errors.js b/lib/errors.js
index e62cee6..757e68a 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var fs = require('fs');
@@ -20,29 +20,7 @@ var verror = require('verror');
 ///--- Globals
 
 var WError = verror.WError;
-
-var slice = Function.prototype.call.bind(Array.prototype.slice);
-
-
-
-///--- Helpers
-
-function ISODateString(d) {
-    function pad(n) {
-        return n < 10 ? '0' + n : n;
-    }
-
-    if (typeof (d) === 'string')
-        d = new Date(d);
-
-    return d.getUTCFullYear() + '-'
-        + pad(d.getUTCMonth()+1) + '-'
-        + pad(d.getUTCDate()) + 'T'
-        + pad(d.getUTCHours()) + ':'
-        + pad(d.getUTCMinutes()) + ':'
-        + pad(d.getUTCSeconds()) + 'Z';
-}
-
+var VError = verror.VError;
 
 
 ///--- Errors
@@ -187,7 +165,7 @@ function InvalidIndexDefinitionError(cause, type) {
                 ['boolean', 'number', 'string'], type);
     this.name = this.constructor.name;
 }
-util.inherits(InvalidIndexTypeError, WError);
+util.inherits(InvalidIndexDefinitionError, WError);
 
 
 function InvalidIndexTypeError(cause, name, type) {
@@ -217,6 +195,13 @@ function InvalidQueryError(cause, filter) {
 util.inherits(InvalidQueryError, WError);
 
 
+function InvocationError() {
+    VError.apply(this, arguments);
+    this.name = this.constructor.name;
+}
+util.inherits(InvocationError, VError);
+
+
 function NoDatabaseError(cause) {
     WError.call(this, (cause || {}),
                 'Not connected to manatee and/or postgres');
@@ -347,6 +332,7 @@ module.exports = {
     InvalidIndexDefinitionError: InvalidIndexDefinitionError,
     InvalidIndexTypeError: InvalidIndexTypeError,
     InvalidQueryError: InvalidQueryError,
+    InvocationError: InvocationError,
     NoDatabaseError: NoDatabaseError,
     NoDatabasePeersError: NoDatabasePeersError,
     NotFunctionError: NotFunctionError,
diff --git a/lib/manatee.js b/lib/manatee.js
index 81b729d..a7263b3 100644
--- a/lib/manatee.js
+++ b/lib/manatee.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var EventEmitter = require('events').EventEmitter;
@@ -27,7 +27,6 @@ var NoDatabasePeersError = require('./errors').NoDatabasePeersError;
 
 ///--- Globals
 
-var slice = Array.prototype.slice;
 var sprintf = util.format;
 
 var DBNAME = process.env.MORAY_DB_NAME || 'moray';
@@ -128,15 +127,10 @@ module.exports = {
 
 ///--- API
 
-Manatee.prototype.pg = function pg(options, callback) {
-    var self = this;
-    if (typeof (options) === 'function') {
-        callback = options;
-        options = {};
-    }
-    assert.object(options, 'options');
+Manatee.prototype.pg = function pg(callback) {
     assert.func(callback, 'callback');
 
+    var self = this;
     var db = self._database;
     var log = self._log;
 
@@ -146,8 +140,7 @@ Manatee.prototype.pg = function pg(options, callback) {
     }
 
     log.debug({
-        db: db,
-        options: options
+        db: db
     }, 'pg: entered');
     db.checkout(function pgCallback(err, client) {
         if (err) {
@@ -164,16 +157,12 @@ Manatee.prototype.pg = function pg(options, callback) {
 };
 
 
-Manatee.prototype.start = function start(opts, cb) {
-    var self = this;
-    if (typeof (opts) === 'function') {
-        cb = opts;
-        opts = {};
-    }
+Manatee.prototype.start = function start(cb) {
+    assert.func(cb, 'cb');
 
-    var log = self._log;
+    var log = this._log;
 
-    self.pg(opts, function (pool_err, client) {
+    this.pg(function (pool_err, client) {
         if (pool_err) {
             cb(pool_err);
             return;
diff --git a/lib/objects/batch.js b/lib/objects/batch.js
index d3f31b9..4e72832 100644
--- a/lib/objects/batch.js
+++ b/lib/objects/batch.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var vasync = require('vasync');
@@ -15,7 +15,9 @@ var clone = require('clone');
 var control = require('../control');
 var common = require('./common');
 var dtrace = require('../dtrace');
-require('../errors');
+
+
+///--- Globals
 
 // Operation Pipelines
 var put = require('./put');
@@ -23,6 +25,16 @@ var update = require('./update');
 var del = require('./del');
 var deleteMany = require('./del_many');
 
+var ARGS_SCHEMA = [
+    { name: 'requests', type: 'requests' },
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    control.getPGHandleAndTransaction,
+    processRequests
+];
+
 
 ///--- Helpers
 
@@ -49,7 +61,7 @@ function processRequests(req, cb) {
                 log: req.log,
                 pg: req.pg,
                 req_id: req.req_id,
-                res: req.res,
+                rpc: req.rpc,
                 bucketCache: req.bucketCache,
                 opts: mergeOptions(r.options, req.opts)
             };
@@ -80,8 +92,13 @@ function processRequests(req, cb) {
             }
 
             dtrace['batch-op-start'].fire(function () {
-                return ([req.msgid, req.req_id, r.bucket, op,
-                        subReq.key || subReq.filter]);
+                return ([
+                    req.msgid,
+                    req.req_id,
+                    r.bucket,
+                    op,
+                    subReq.key || subReq.filter
+                ]);
             });
 
             vasync.pipeline({
@@ -97,8 +114,8 @@ function processRequests(req, cb) {
                 } else {
                     var out = {
                         bucket: r.bucket,
-                        count: subReq.res._count,
-                        etag: subReq.res._etag
+                        count: subReq._count,
+                        etag: subReq._etag
                     };
                     if (r.key) {
                         out.key = r.key;
@@ -124,27 +141,30 @@ function processRequests(req, cb) {
 
 function batch(options) {
     control.assertOptions(options);
-    var route = 'batch';
 
-    function _batch(requests, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _batch(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var requests = argv[0];
+        var opts = argv[1];
+
+        var req = control.buildReq(opts, rpc, options);
         req.requests = requests;
 
         dtrace['batch-start'].fire(function () {
-            return ([res.msgid, req.req_id]);
+            return ([req.msgid, req.req_id]);
         });
         req.log.debug({
             requests: requests,
             opts: opts
-        }, '%s: entered', route);
+        }, 'batch: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true),
-                processRequests
-            ],
+            funcs: PIPELINE,
             cbOutput: function () { return {etags: req.etags}; }
         });
     }
diff --git a/lib/objects/common.js b/lib/objects/common.js
index 185f136..85bab9c 100644
--- a/lib/objects/common.js
+++ b/lib/objects/common.js
@@ -686,7 +686,7 @@ function checkEtag(req, cb) {
 function stdOutput(req) {
     // Default output (used by all but batch and findObjects)
     return (function () {
-        return { count: req.res._count, etag: req.res._etag };
+        return { count: req._count, etag: req._etag };
     });
 }
 
diff --git a/lib/objects/del.js b/lib/objects/del.js
index 73373e8..5a652d9 100644
--- a/lib/objects/del.js
+++ b/lib/objects/del.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -23,13 +23,22 @@ var EtagConflictError = mod_errors.EtagConflictError;
 
 ///--- Globals
 
-// This is exported so that batch put can leverage it
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'key', type: 'string' },
+    { name: 'options', type: 'options' }
+];
+
 var PIPELINE = [
+    control.getPGHandleAndTransaction,
     common.loadBucket,
     drop,
     common.runPostChain
 ];
 
+// This is exported for batch 'delete' operations
+var SUBPIPELINE = PIPELINE.slice(1);
+
 ///--- Handlers
 
 function drop(req, cb) {
@@ -80,10 +89,18 @@ function drop(req, cb) {
 
 function del(options) {
     control.assertOptions(options);
-    var route = 'delObject';
 
-    function _del(bucket, key, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _del(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var bucket = argv[0];
+        var key = argv[1];
+        var opts = argv[2];
+
+        var req = control.buildReq(opts, rpc, options);
         req.bucket = {
             name: bucket
         };
@@ -91,7 +108,7 @@ function del(options) {
         req.etag = (opts.etag || opts._etag);
 
         dtrace['delobject-start'].fire(function () {
-            return ([res.msgid, req.req_id, bucket, key]);
+            return ([req.msgid, req.req_id, bucket, key]);
         });
         req.log.debug({
             bucket: bucket,
@@ -100,11 +117,8 @@ function del(options) {
         }, 'delObject: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true)
-            ].concat(PIPELINE),
+            funcs: PIPELINE,
             cbOutput: common.stdOutput(req)
         });
     }
@@ -118,5 +132,5 @@ function del(options) {
 
 module.exports = {
     del: del,
-    pipeline: PIPELINE
+    pipeline: SUBPIPELINE
 };
diff --git a/lib/objects/del_many.js b/lib/objects/del_many.js
index a5053a1..bb6c9bf 100644
--- a/lib/objects/del_many.js
+++ b/lib/objects/del_many.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -15,13 +15,18 @@ var once = require('once');
 var control = require('../control');
 var common = require('./common');
 var dtrace = require('../dtrace');
-require('../errors');
 
 
 ///--- Globals
 
-// This is exported so that batch put can leverage it
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'filter', type: 'string' },
+    { name: 'options', type: 'options' }
+];
+
 var PIPELINE = [
+    control.getPGHandleAndTransaction,
     common.parseFilter,
     common.loadBucket,
     common.decorateFilter,
@@ -29,6 +34,9 @@ var PIPELINE = [
     drop
 ];
 
+// This is exported for batch 'deleteMany' operations
+var SUBPIPELINE = PIPELINE.slice(1);
+
 
 ///--- Handlers
 
@@ -60,7 +68,7 @@ function drop(req, cb) {
             id: req._id,
             res: res
         }, 'drop: done');
-        req.res._count = res.rowCount;
+        req._count = res.rowCount;
         cb();
     });
 }
@@ -68,30 +76,35 @@ function drop(req, cb) {
 
 function deleteMany(options) {
     control.assertOptions(options);
-    var route = 'deleteMany';
 
-    function _deleteMany(bucket, filter, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _deleteMany(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var bucket = argv[0];
+        var filter = argv[1];
+        var opts = argv[2];
+
+        var req = control.buildReq(opts, rpc, options);
         req.bucket = {
             name: bucket
         };
         req.rawFilter = filter;
 
         dtrace['delmany-start'].fire(function () {
-            return ([res.msgid, req.req_id, bucket, filter]);
+            return ([req.msgid, req.req_id, bucket, filter]);
         });
         req.log.debug({
             bucket: bucket,
             filter: filter,
             opts: opts
-        }, '%s: entered', route);
+        }, 'deleteMany: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true)
-            ].concat(PIPELINE),
+            funcs: PIPELINE,
             cbOutput: common.stdOutput(req)
         });
     }
@@ -104,5 +117,5 @@ function deleteMany(options) {
 
 module.exports = {
     deleteMany: deleteMany,
-    pipeline: PIPELINE
+    pipeline: SUBPIPELINE
 };
diff --git a/lib/objects/find.js b/lib/objects/find.js
index 114c8bb..d706e09 100644
--- a/lib/objects/find.js
+++ b/lib/objects/find.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
@@ -16,6 +16,28 @@ var common = require('./common');
 var dtrace = require('../dtrace');
 var errors = require('../errors');
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'filter', type: 'string' },
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    common.parseFilter,
+    control.getPGHandle,
+    beginRepeatableRead,
+    common.loadBucket,
+    checkRequiredIndexes,
+    sendHandledOptions,
+    common.decorateFilter,
+    common.buildWhereClause,
+    getCount,
+    getRecords,
+    sendSQL
+];
+
 var HANDLED_FINDOBJECTS_OPTIONS = [
     'req_id',
     'limit',
@@ -142,7 +164,6 @@ function getRecords(req, cb) {
     var bucket = req.bucket.name;
     var filter = req.filter;
     var log = req.log;
-    var res = req.res;
     var sql = util.format('SELECT *, \'%s\' AS req_id FROM %s %s',
                           req.req_id, bucket, req.where.clause);
 
@@ -221,15 +242,17 @@ function getRecords(req, cb) {
             delete v._key;
             delete v._mtime;
             delete v._count;
-            res.write(obj);
+            req.rpc.write(obj);
             dtrace['findobjects-record'].fire(function () {
-                return ([res.msgid,
-                         obj.key,
-                         obj._id,
-                         obj._etag,
-                         row._value]);
+                return ([
+                    req.msgid,
+                    obj.key,
+                    obj._id,
+                    obj._etag,
+                    row._value
+                ]);
             });
-            res._num_records++;
+            req._num_records++;
         }
     });
 
@@ -243,53 +266,45 @@ function getRecords(req, cb) {
 
 function sendSQL(req, cb) {
     if (req.opts.sql_only) {
-        req.res.write(req.opts._sql);
+        req.rpc.write(req.opts._sql);
     }
     cb(null);
 }
 
 function find(options) {
     control.assertOptions(options);
-    var route = 'findObjects';
 
-    function _find(bucket, filter, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _find(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var bucket = argv[0];
+        var filter = argv[1];
+        var opts = argv[2];
+
+        var req = control.buildReq(opts, rpc, options);
         req.bucket = {
             name: bucket
         };
         req.rawFilter = filter;
-        req.res._num_records = 0;
+        req._num_records = 0;
 
         dtrace['findobjects-start'].fire(function () {
-            return ([res.msgid, req.req_id, bucket, filter]);
+            return ([req.msgid, req.req_id, bucket, filter]);
         });
         req.log.debug({
             bucket: bucket,
             filter: filter,
             opts: opts
-        }, '%s: entered', route);
+        }, 'findObjects: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                common.parseFilter,
-                control.getPGHandle(route, options.manatee, {
-                    async: false,
-                    read: true
-                }),
-                beginRepeatableRead,
-                common.loadBucket,
-                checkRequiredIndexes,
-                sendHandledOptions,
-                common.decorateFilter,
-                common.buildWhereClause,
-                getCount,
-                getRecords,
-                sendSQL
-            ],
+            funcs: PIPELINE,
             cbOutput: function () { return; },
-            cbProbe: function () { return [res.msgid, res._num_records]; }
+            cbProbe: function () { return [req.msgid, req._num_records]; }
         });
     }
 
diff --git a/lib/objects/get.js b/lib/objects/get.js
index 724d667..c0aa9f2 100644
--- a/lib/objects/get.js
+++ b/lib/objects/get.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -20,6 +20,34 @@ var dtrace = require('../dtrace');
 var ObjectNotFoundError = require('../errors').ObjectNotFoundError;
 
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'key', type: 'string' },
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    checkCache,
+    function dbConnect(r, cb) {
+        if (r.object) {
+            cb(null);
+        } else {
+            control.getPGHandle(r, cb);
+        }
+    },
+    function loadBucket(r, cb) {
+        if (r.object) {
+            cb();
+        } else {
+            common.loadBucket(r, cb);
+        }
+    },
+    loadObject
+];
+
+
 ///--- Handlers
 
 function checkCache(req, cb) {
@@ -87,10 +115,18 @@ function loadObject(req, cb) {
 function get(options) {
     control.assertOptions(options);
     assert.object(options.objectCache);
-    var route = 'getObject';
 
-    function _get(bucket, key, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _get(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var bucket = argv[0];
+        var key = argv[1];
+        var opts = argv[2];
+
+        var req = control.buildReq(opts, rpc, options);
         req.bucket = {
             name: bucket
         };
@@ -98,36 +134,17 @@ function get(options) {
         req.key = key;
 
         dtrace['getobject-start'].fire(function () {
-            return ([res.msgid, req.req_id, bucket, key]);
+            return ([req.msgid, req.req_id, bucket, key]);
         });
         req.log.debug({
             bucket: bucket,
             key: key,
             opts: opts
-        }, '%s: entered', route);
+        }, 'getObject: entered');
+
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                checkCache,
-                function dbConnect(r, cb) {
-                    if (r.object) {
-                        cb(null);
-                    } else {
-                        var fetch = control.getPGHandle(route, options.manatee,
-                                {async: false, read: true});
-                        fetch(r, cb);
-                    }
-                },
-                function loadBucket(r, cb) {
-                    if (r.object) {
-                        cb();
-                    } else {
-                        common.loadBucket(r, cb);
-                    }
-                },
-                loadObject
-            ],
+            funcs: PIPELINE,
             cbOutput: function () { return req.object; },
             cbProbe: function () {
                 var val = JSON.stringify(req.object);
diff --git a/lib/objects/put.js b/lib/objects/put.js
index 65742f7..fd1e54e 100644
--- a/lib/objects/put.js
+++ b/lib/objects/put.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -17,14 +17,19 @@ var vasync = require('vasync');
 var control = require('../control');
 var common = require('./common');
 var dtrace = require('../dtrace');
-require('../errors');
-
 
 
 ///--- Globals
 
-// This is exported so that batch put can leverage it
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'key', type: 'string' },
+    { name: 'value', type: 'object' },
+    { name: 'options', type: 'options' }
+];
+
 var PIPELINE = [
+    control.getPGHandleAndTransaction,
     common.loadBucket,
     common.selectForUpdate,
     common.verifyBucket,
@@ -38,6 +43,8 @@ var PIPELINE = [
     common.runPostChain
 ];
 
+// This is exported fo batch 'put' operations
+var SUBPIPELINE = PIPELINE.slice(1);
 
 
 ///--- Internal Functions
@@ -178,9 +185,9 @@ function insert(req, cb) {
     var q;
     var sql;
 
-    req.res._etag = _createEtag(req);
+    req._etag = _createEtag(req);
     var fields = ['_key', '_value', '_etag', '_mtime', '_vnode'];
-    var values = [req.key, req._value, req.res._etag, _now(), req.value.vnode];
+    var values = [req.key, req._value, req._etag, _now(), req.value.vnode];
     if (req.bucket.options.guaranteeOrder === true) {
         fields.push('_txn_snap');
         values.push(req.value._txn_snap);
@@ -243,10 +250,10 @@ function update(req, cb) {
     var track = (req.bucket.options.trackModification === true);
     var _txn_snap = req.value._txn_snap;
 
-    req.res._etag = _createEtag(req);
+    req._etag = _createEtag(req);
 
     var fields = ['_value', '_etag', '_vnode', '_mtime'];
-    var values = [req._value, req.res._etag, req.value.vnode, _now()];
+    var values = [req._value, req._etag, req.value.vnode, _now()];
     if (req.bucket.reindex_active) {
         /* update _rver if table undergoing reindexing */
         fields.push('_rver');
@@ -332,10 +339,19 @@ function updateSerial(req, cb) {
 
 function put(options) {
     control.assertOptions(options);
-    var route = 'putObject';
 
-    function _put(bucket, key, value, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _put(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var bucket = argv[0];
+        var key = argv[1];
+        var value = argv[2];
+        var opts = argv[3];
+
+        var req = control.buildReq(opts, rpc, options);
         req.bucket = {
             name: bucket
         };
@@ -345,7 +361,7 @@ function put(options) {
         req.etag = opts.etag;
 
         dtrace['putobject-start'].fire(function () {
-            return ([res.msgid, req.req_id, bucket, key, opts._value]);
+            return ([req.msgid, req.req_id, bucket, key, opts._value]);
         });
         req.log.debug({
             bucket: bucket,
@@ -355,11 +371,8 @@ function put(options) {
         }, 'putObject: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true)
-            ].concat(PIPELINE),
+            funcs: PIPELINE,
             cbOutput: common.stdOutput(req)
         });
     }
@@ -372,5 +385,5 @@ function put(options) {
 
 module.exports = {
     put: put,
-    pipeline: PIPELINE
+    pipeline: SUBPIPELINE
 };
diff --git a/lib/objects/reindex.js b/lib/objects/reindex.js
index 7958f96..58ca424 100644
--- a/lib/objects/reindex.js
+++ b/lib/objects/reindex.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -17,8 +17,23 @@ var vasync = require('vasync');
 var control = require('../control');
 var common = require('./common');
 var dtrace = require('../dtrace');
-require('../errors');
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'count', type: 'integer' },
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    control.getPGHandleAndTransaction,
+    common.loadBucket,
+    requiresReindex,
+    processRows,
+    recordStatus,
+    countRemaining
+];
 
 ///--- Handlers
 
@@ -225,10 +240,18 @@ function countRemaining(req, cb) {
 
 function reindex(options) {
     control.assertOptions(options);
-    var route = 'reindexObjects';
 
-    function _reindex(bucket, count, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _reindex(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var bucket = argv[0];
+        var count = argv[1];
+        var opts = argv[2];
+
+        var req = control.buildReq(opts, rpc, options);
         req.count = parseInt(count, 10);
         req.bucket = {
             name: bucket
@@ -237,25 +260,17 @@ function reindex(options) {
         req.opts.noBucketCache = true;
 
         dtrace['reindexobjects-start'].fire(function () {
-            return ([res.msgid, req.req_id, bucket, count]);
+            return ([req.msgid, req.req_id, bucket, count]);
         });
         req.log.debug({
             bucket: bucket,
             count: count,
             opts: opts
-        }, '%s: entered', route);
+        }, 'reindexObjects: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true),
-                common.loadBucket,
-                requiresReindex,
-                processRows,
-                recordStatus,
-                countRemaining
-            ],
+            funcs: PIPELINE,
             cbOutput: function () {
                 var result = { processed: req._processed };
                 if (req._countRemaining !== undefined) {
diff --git a/lib/objects/update.js b/lib/objects/update.js
index 13810b9..8f60178 100644
--- a/lib/objects/update.js
+++ b/lib/objects/update.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -24,8 +24,15 @@ var FieldUpdateError = mod_errors.FieldUpdateError;
 
 ///--- Globals
 
-// This is exported so that batch put can leverage it
+var ARGS_SCHEMA = [
+    { name: 'bucket', type: 'string' },
+    { name: 'fields', type: 'object' },
+    { name: 'filter', type: 'string' },
+    { name: 'options', type: 'options' }
+];
+
 var PIPELINE = [
+    control.getPGHandleAndTransaction,
     common.parseFilter,
     common.loadBucket,
     common.decorateFilter,
@@ -33,6 +40,8 @@ var PIPELINE = [
     updateRows
 ];
 
+// This is exported for batch 'update' operations
+var SUBPIPELINE = PIPELINE.slice(1);
 
 ///--- Handlers
 
@@ -98,11 +107,11 @@ function updateRows(req, cb) {
         cb(err);
     });
     q.once('end', function (res) {
-        req.res._etag = etag;
-        req.res._count = res.rowCount;
+        req._etag = etag;
+        req._count = res.rowCount;
         log.debug({
-            count: req.res._count,
-            etag: req.res._etag
+            count: req._count,
+            etag: req._etag
         }, 'updateRows: done');
         cb();
     });
@@ -111,37 +120,43 @@ function updateRows(req, cb) {
 
 function update(options) {
     control.assertOptions(options);
-    var route = 'updateObjects';
 
-    function _update(bucket, fields, filter, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _update(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var bucket = argv[0];
+        var fields = argv[1];
+        var filter = argv[2];
+        var opts = argv[3];
+
+        var req = control.buildReq(opts, rpc, options);
         req.bucket = {
             name: bucket
         };
         req.fields = fields;
         req.fieldKeys = Object.keys(fields);
         if (req.fieldKeys.length === 0) {
-            res.end(new FieldUpdateError(fields));
+            rpc.fail(new FieldUpdateError(fields));
             return;
         }
         req.rawFilter = filter;
 
         dtrace['update-start'].fire(function () {
-            return ([res.msgid, req.req_id, bucket, fields, filter]);
+            return ([req.msgid, req.req_id, bucket, fields, filter]);
         });
         req.log.debug({
             bucket: bucket,
             fields: fields,
             filter: filter,
             opts: opts
-        }, '%s: entered', route);
+        }, 'updateObjects: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true)
-            ].concat(PIPELINE),
+            funcs: PIPELINE,
             cbOutput: common.stdOutput(req)
         });
     }
@@ -154,5 +169,5 @@ function update(options) {
 
 module.exports = {
     update: update,
-    pipeline: PIPELINE
+    pipeline: SUBPIPELINE
 };
diff --git a/lib/pg.js b/lib/pg.js
index 07ab4d4..796bbd9 100644
--- a/lib/pg.js
+++ b/lib/pg.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016, Joyent, Inc.
+ * Copyright 2017, Joyent, Inc.
  */
 
 var crypto = require('crypto');
@@ -33,9 +33,6 @@ var UniqueAttributeError = mod_errors.UniqueAttributeError;
 
 ///--- Globals
 
-var slice = Array.prototype.slice;
-var sprintf = util.format;
-
 var CLIENT_ID = 0;
 
 var SERIALIZERS = {
@@ -460,12 +457,7 @@ function PGPool(options) {
     this.url = options.url;
 
     function reEmit(event) {
-        if (self.log.debug()) {
-            var args = slice(arguments);
-            args.unshift();
-            self.log.debug('pool event %s: %j', event, args);
-        }
-
+        self.log.debug('pool event: %s', event);
         self.emit.apply(self, arguments);
     }
 
diff --git a/lib/ping.js b/lib/ping.js
index 7005d1d..dc292ab 100644
--- a/lib/ping.js
+++ b/lib/ping.js
@@ -5,15 +5,22 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
 
 var assert = require('assert-plus');
+var control = require('./control');
 var libuuid = require('libuuid');
 
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'options', type: 'options' }
+];
+
 
 ///--- Handlers
 
@@ -25,7 +32,13 @@ function ping(options) {
 
     var manatee = options.manatee;
 
-    function _ping(opts, res) {
+    function _ping(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var opts = argv[0];
         var log = options.log.child({
             req_id: opts.req_id || libuuid.create()
         });
@@ -36,14 +49,14 @@ function ping(options) {
 
         if (!opts.deep) {
             log.debug('ping: done');
-            res.end();
+            rpc.end();
             return;
         }
 
         manatee.pg(function (startErr, pg) {
             if (startErr) {
                 log.debug(startErr, 'ping: no DB handle');
-                res.end(startErr);
+                rpc.fail(startErr);
                 return;
             }
 
@@ -55,7 +68,7 @@ function ping(options) {
                 req.removeAllListeners('end');
                 req.removeAllListeners('row');
                 pg.release();
-                res.end(err);
+                rpc.fail(err);
             });
 
             req.once('row', function (r) {
@@ -67,7 +80,7 @@ function ping(options) {
                 req.removeAllListeners('error');
                 req.removeAllListeners('row');
                 pg.release();
-                res.end();
+                rpc.end();
             });
         });
 
@@ -81,16 +94,23 @@ function version(options, ver) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
 
-    function _version(opts, res) {
+    function _version(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var opts = argv[0];
         var log = options.log.child({
             req_id: opts.req_id || libuuid.create()
         });
         log.debug({
             opts: opts
         }, 'version: entered');
-        res.end({version: ver});
-        log.debug('request done');
+        rpc.end({ version: ver });
+        log.debug('version: done');
     }
+
     return (_version);
 }
 
diff --git a/lib/schema.js b/lib/schema.js
new file mode 100644
index 0000000..05f7662
--- /dev/null
+++ b/lib/schema.js
@@ -0,0 +1,331 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2017, Joyent, Inc.
+ */
+
+/*
+ * This file contains schemas used for validating arguments to Moray
+ * endpoints, such as base types, types of provided options, bucket
+ * configurations, etc.
+ *
+ * These schemas are defined using JSON schema version 4.
+ */
+
+var Ajv = require('ajv');
+var assert = require('assert-plus');
+
+var mod_errors = require('./errors');
+var InvocationError = mod_errors.InvocationError;
+var InvalidBucketConfigError = mod_errors.InvalidBucketConfigError;
+
+///--- Globals
+
+var AJV_ENV = new Ajv();
+var MORAY_TYPES = [
+    'string', '[string]',
+    'number', '[number]',
+    'boolean', '[boolean]',
+    'ip', '[ip]',
+    'subnet', '[subnet]'
+];
+
+
+///--- Internal helpers
+
+function allowedValuesText(allowed) {
+    assert.array(allowed, 'allowed');
+
+    var seen = {};
+    var values = [];
+
+    allowed.map(JSON.stringify).forEach(function (str) {
+        var key = str.toUpperCase();
+        if (!seen.hasOwnProperty(key)) {
+            seen[key] = true;
+            values.push(str);
+        }
+    });
+
+    return ' (' + values.join(', ') + ')';
+}
+
+
+function errorMessage(err, name) {
+    var msg = name + err.dataPath + ' ' + err.message;
+
+    if (err.params.hasOwnProperty('allowedValues')) {
+        msg += allowedValuesText(err.params.allowedValues);
+    }
+
+    return msg;
+}
+
+
+function errorsText(errs, name) {
+    return errs.map(function (err) {
+        return errorMessage(err, name);
+    }).join(', ');
+}
+
+
+///--- Schema Declarations
+
+AJV_ENV.addSchema({
+    id: 'integer',
+    type: 'integer',
+    minimum: 0
+});
+
+AJV_ENV.addSchema({
+    id: 'string',
+    type: 'string',
+    minLength: 1
+});
+
+AJV_ENV.addSchema({
+    id: 'array',
+    type: 'array'
+});
+
+AJV_ENV.addSchema({
+    id: 'object',
+    type: 'object',
+    properties: {
+        'hasOwnProperty': {
+            description: '"hasOwnProperty" should not be on received objects',
+            not: {
+                type: [
+                    'number',
+                    'string',
+                    'boolean',
+                    'array',
+                    'object',
+                    'null'
+                ]
+            }
+        }
+    }
+});
+
+AJV_ENV.addSchema({
+    id: 'sortobj',
+    type: 'object',
+    allOf: [ { '$ref': 'object' } ],
+    properties: {
+        attribute: {
+            description: 'Column to sort on',
+            type: 'string'
+        },
+        order: {
+            description: 'Direction in which to sort',
+            type: 'string',
+            enum: [ 'ASC', 'asc', 'DESC', 'desc' ]
+        }
+    }
+});
+
+AJV_ENV.addSchema({
+    id: 'options',
+    type: 'object',
+    allOf: [ { '$ref': 'object' } ],
+    properties: {
+        // Common options
+        'timeout': {
+            description: 'How long to wait for a response to SQL queries',
+            type: 'integer',
+            minimum: 0
+        },
+        'req_id': {
+            description: 'A unique identifier for this request',
+            type: 'string',
+            format: 'uuid',
+            maxLength: 36
+        },
+
+        // Object RPC options
+        'etag': {
+            description: 'The expected, current etag of the targeted object',
+            type: [ 'null', 'string' ]
+        },
+        '_value': {
+            description: 'An stringified version of the sent object',
+            type: 'string'
+        },
+        'noBucketCache': {
+            description: 'Whether to check the bucket cache for bucket schemas',
+            type: 'boolean'
+        },
+        'noCache': {
+            description: 'Whether to check the object cache before Postgres',
+            type: 'boolean'
+        },
+        'headers': {
+            description: 'Information to pass to pre/post-triggers',
+            allOf: [ { '$ref': 'object' } ]
+        },
+
+        // Filter options
+        'noLimit': {
+            description: 'Skip using the default result limit',
+            type: 'boolean'
+        },
+        'limit': {
+            description: 'The maximum number of objects to return for a query',
+            type: [ 'integer', 'string' ],
+            pattern: '^[0-9]+$',
+            minimum: 0
+        },
+        'offset': {
+            description: 'An offset into a query result set',
+            type: 'number',
+            minimum: 0
+        },
+        'sort': {
+            description: 'Sort the results of a query',
+            anyOf: [
+                { type: 'array', items: { '$ref': 'sortobj' } },
+                { '$ref': 'sortobj' }
+            ]
+        },
+        'no_count': {
+            description: 'Skip calculating how many rows match the query',
+            type: 'boolean'
+        },
+        'sql_only': {
+            description: 'Return the SQL that would be executed ' +
+                'instead of running it',
+            type: 'boolean'
+        },
+
+        // Misc options
+        'no_reindex': {
+            type: 'boolean'
+        },
+        'deep': {
+            type: 'boolean'
+        }
+    }
+});
+
+AJV_ENV.addSchema({
+    id: 'requestobj',
+    type: 'object',
+    allOf: [ { '$ref': 'object' } ],
+    required: [ 'bucket' ],
+    properties: {
+        operation: {
+            description: 'The operation this request will perform',
+            type: 'string',
+            enum: [ 'put', 'delete', 'update', 'deleteMany' ]
+        },
+        bucket: {
+            description: 'The bucket this request should operate on',
+            type: 'string'
+        },
+        key: {
+            description: 'The key in the bucket to affect',
+            type: 'string'
+        },
+        filter: {
+            description: 'A query describing which objects to affect',
+            type: 'string'
+        },
+        fields: {
+            description: 'The fields to update and their new values',
+            allOf: [ { '$ref': 'object' } ]
+        },
+        value: {
+            description: 'The value being inserted into the bucket',
+            allOf: [ { '$ref': 'object' } ]
+        },
+        options: { '$ref': 'options' }
+    }
+});
+
+
+AJV_ENV.addSchema({
+    id: 'requests',
+    type: 'array',
+    items: { '$ref': 'requestobj' }
+});
+
+
+AJV_ENV.addSchema({
+    id: 'index',
+    type: 'object',
+    allOf: [ { '$ref': 'object' } ],
+    required: [ 'type' ],
+    additionalProperties: false,
+    properties: {
+        'type': {
+            type: 'string',
+            enum: MORAY_TYPES
+        },
+        'unique': {
+            type: 'boolean'
+        }
+    }
+});
+
+AJV_ENV.addSchema({
+    id: 'bucket',
+    type: 'object',
+    allOf: [ { '$ref': 'object' } ],
+    properties: {
+        'index': {
+            allOf: [ { '$ref': 'object' } ],
+            patternProperties: {
+                '.*': { '$ref': 'index' }
+            }
+        },
+        'pre': {
+            type: 'array'
+        },
+        'post': {
+            type: 'array'
+        },
+        'options': {
+            allOf: [ { '$ref': 'object' } ],
+            properties: {
+                'version': { '$ref': 'integer' },
+                'trackModification': {
+                    type: 'boolean'
+                },
+                'guaranteeOrder': {
+                    type: 'boolean'
+                },
+                'syncUpdates': {
+                    type: 'boolean'
+                }
+            }
+        }
+    }
+});
+
+
+///--- Exported functions
+
+exports.validateArgument = function validateArgument(name, schema, value) {
+    assert.string(name, 'name');
+    assert.string(schema, 'schema name');
+
+    if (AJV_ENV.validate(schema, value)) {
+        return null;
+    }
+
+    return new InvocationError('%s', errorsText(AJV_ENV.errors, name));
+};
+
+exports.validateBucket = function validateBucket(bucket) {
+    if (AJV_ENV.validate('bucket', bucket)) {
+        return null;
+    }
+
+    return new InvalidBucketConfigError('%s',
+        errorsText(AJV_ENV.errors, 'bucket'));
+};
diff --git a/lib/server.js b/lib/server.js
index 06d51bf..ba8e49c 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -5,18 +5,23 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
 var EventEmitter = require('events').EventEmitter;
 
 var assert = require('assert-plus');
-var fast = require('fast');
 var libuuid = require('libuuid');
+var mod_fast = require('fast');
+var mod_kang = require('kang');
+var mod_net = require('net');
+var mod_os = require('os');
 var VError = require('verror').VError;
 var LRU = require('lru-cache');
+var vasync = require('vasync');
 
+var control = require('./control');
 var buckets = require('./buckets');
 var objects = require('./objects');
 var ping = require('./ping');
@@ -25,14 +30,30 @@ var sql = require('./sql');
 
 ///--- Globals
 
-var slice = Function.prototype.call.bind(Array.prototype.slice);
+/*
+ * API version, increment for depended-upon changes.
+ *
+ * This should not be incremented lightly: Moray consumers don't currently
+ * have a great way to deal with Moray backends of different versions. Work
+ * will need to be done there before this can be bumped. See RFD 33 for a
+ * longer discussion on this issue.
+ */
+var API_VERSION = 2;
 
+/*
+ * This is the Kang version identifier. For now, we set it to 1.0.0. Someday,
+ * we'll hopefully have a better way of grabbing information about the image
+ * version and exporting that here.
+ */
+var KANG_VERSION = '1.0.0';
 
 
 ///--- API
 
 
 function MorayServer(options) {
+    var self = this;
+
     EventEmitter.call(this);
 
     var log = options.log;
@@ -61,84 +82,143 @@ function MorayServer(options) {
         bucketCache: this.ms_bucketCache,
         objectCache: this.ms_objectCache
     };
-    var server = fast.createServer({log: log});
-
-    server.rpc('createBucket', buckets.creat(opts));
-    server.rpc('getBucket', buckets.get(opts));
-    server.rpc('listBuckets', buckets.list(opts));
-    server.rpc('updateBucket', buckets.update(opts));
-    server.rpc('delBucket', buckets.del(opts));
-    server.rpc('putObject', objects.put(opts));
-    server.rpc('batch', objects.batch(opts));
-    server.rpc('getObject', objects.get(opts));
-    server.rpc('delObject', objects.del(opts));
-    server.rpc('findObjects', objects.find(opts));
-    server.rpc('updateObjects', objects.update(opts));
-    server.rpc('reindexObjects', objects.reindex(opts));
-    server.rpc('deleteMany', objects.deleteMany(opts));
-    server.rpc('getTokens', getTokens(opts));
-    server.rpc('sql', sql.sql(opts));
-    server.rpc('ping', ping.ping(opts));
-    // API version, increment for depended-upon changes
-    var API_VERSION = 2;
-    server.rpc('version', ping.version(opts, API_VERSION));
-
-    if (options.audit !== false) {
-        server.on('after', function (name, req, res) {
-            var t = Math.floor(res.elapsed / 1000);
-            var obj = {
-                method: name,
-                'arguments': req,
-                serverTime: t + 'ms'
-            };
-
-            log.info(obj, 'request handled');
-        });
-    }
 
-    server.on('error', function (err) {
-        log.error(err, 'server error');
-        throw new VError(err, 'unsolicited server error');
+    var socket = mod_net.createServer({ 'allowHalfOpen': true });
+    var server = new mod_fast.FastServer({
+        log: log.child({ component: 'fast' }),
+        server: socket
     });
-    server.on('uncaughtException', function (err) {
-        log.error(err, 'uncaught: server error');
-        var args = slice(arguments);
-        args.shift();
-        args[args.length - 1].end(err);
+
+    var methods = [
+        { rpcmethod: 'createBucket', rpchandler: buckets.creat(opts) },
+        { rpcmethod: 'getBucket', rpchandler: buckets.get(opts) },
+        { rpcmethod: 'listBuckets', rpchandler: buckets.list(opts) },
+        { rpcmethod: 'updateBucket', rpchandler: buckets.update(opts) },
+        { rpcmethod: 'delBucket', rpchandler: buckets.del(opts) },
+        { rpcmethod: 'putObject', rpchandler: objects.put(opts) },
+        { rpcmethod: 'batch', rpchandler: objects.batch(opts) },
+        { rpcmethod: 'getObject', rpchandler: objects.get(opts) },
+        { rpcmethod: 'delObject', rpchandler: objects.del(opts) },
+        { rpcmethod: 'findObjects', rpchandler: objects.find(opts) },
+        { rpcmethod: 'updateObjects', rpchandler: objects.update(opts) },
+        { rpcmethod: 'reindexObjects', rpchandler: objects.reindex(opts) },
+        { rpcmethod: 'deleteMany', rpchandler: objects.deleteMany(opts) },
+        { rpcmethod: 'getTokens', rpchandler: getTokens(opts) },
+        { rpcmethod: 'sql', rpchandler: sql.sql(opts) },
+        { rpcmethod: 'ping', rpchandler: ping.ping(opts) },
+        { rpcmethod: 'version', rpchandler: ping.version(opts, API_VERSION) }
+    ];
+
+    methods.forEach(function (rpc) {
+        server.registerRpcMethod(rpc);
     });
 
     this.port = options.port;
     this.ip = options.bindip;
 
+    this.fast_socket = socket;
     this.fast_server = server;
+    this.kang_server = null;
     this.db_conn = db;
     this.log = options.log;
 
-    var self = this;
-    ['listening', 'error'].forEach(function (event) {
-        // re-emit certain events from fast server
-        self.fast_server.on(event, self.emit.bind(self, event));
+    if (options.kangPort) {
+        mod_kang.knStartServer({
+            port: options.kangPort,
+            host: options.bindip,
+            uri_base: '/kang',
+            service_name: 'moray',
+            version: KANG_VERSION,
+            ident: mod_os.hostname() + '/' + process.pid,
+            list_types: server.kangListTypes.bind(server),
+            list_objects: server.kangListObjects.bind(server),
+            get: server.kangGetObject.bind(server),
+            stats: server.kangStats.bind(server)
+        }, function (err, kang) {
+            if (err) {
+                self.log.error(err, 'failed to start Kang');
+            } else {
+                self.kang_server = kang;
+            }
+        });
+    }
+
+
+    /*
+     * Once both the server socket and the database connection are up,
+     * emit the "ready" event to indicate that the database is good to
+     * use now.
+     *
+     * Failure to listen() will be re-emitted on the MorayServer. In a
+     * "moray" zone, no one is listening, so the error will be thrown,
+     * and SMF will either restart us or take the service into
+     * maintenance.
+     */
+    var sock_ready = false;
+    var dbc_ready = false;
+
+    function emitReady() {
+        if (sock_ready && dbc_ready) {
+            self.emit('ready');
+        }
+    }
+
+    self.fast_socket.on('listening', function () {
+        self.log.info({ address: self.fast_socket.address() },
+            'Moray listening on port %d', self.port);
+        sock_ready = true;
+        emitReady();
     });
-    ['ready'].forEach(function (event) {
-        // re-emit certain events from manatee
-        self.db_conn.on(event, self.emit.bind(self, event));
+
+    self.fast_socket.on('error', function (err) {
+        self.fast_socket = null;
+        self.emit('error', err);
+    });
+
+    self.db_conn.on('ready', function () {
+        dbc_ready = true;
+        emitReady();
     });
 }
 util.inherits(MorayServer, EventEmitter);
 
 
 MorayServer.prototype.listen = function listen() {
-    var self = this;
-    this.fast_server.listen(this.port, this.ip, function () {
-        self.log.info('moray listening on %d', self.port);
-    });
+    this.fast_socket.listen(this.port, this.ip);
 };
 
 
 MorayServer.prototype.close = function close() {
-    this.fast_server.close();
-    this.db_conn.on('close', this.emit.bind(this, 'close'));
-    this.db_conn.close();
+    var barrier = vasync.barrier();
+    var self = this;
+
+    /*
+     * Wait until both the server socket and the DB connection
+     * have been shutdown before closing Fast and Kang.
+     */
+    barrier.on('drain', function () {
+        self.fast_server.close();
+
+        if (self.kang_server !== null) {
+            self.kang_server.close();
+        }
+
+        self.emit('close');
+    });
+
+    if (self.fast_socket !== null) {
+        barrier.start('close-server-socket');
+        self.fast_socket.on('close', function () {
+            barrier.done('close-server-socket');
+        });
+        self.fast_socket.close();
+    }
+
+    barrier.start('close-db-conn');
+    self.db_conn.on('close', function () {
+        barrier.done('close-db-conn');
+    });
+    self.db_conn.close();
 };
 
 
@@ -161,11 +241,21 @@ module.exports = {
 
 ///--- Privates
 
+var GET_TOKENS_ARGS = [
+    { name: 'options', type: 'options' }
+];
+
 function getTokens(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
 
-    function _getTokens(opts, res) {
+    function _getTokens(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, GET_TOKENS_ARGS)) {
+            return;
+        }
+
+        var opts = argv[0];
         var id = opts.req_id || libuuid.create();
         var log = options.log.child({
             req_id: id
@@ -173,7 +263,7 @@ function getTokens(options) {
 
         log.debug('getTokens: entered');
 
-        res.end(new Error('Operation not supported'));
+        rpc.fail(new Error('Operation not supported'));
     }
 
     return _getTokens;
diff --git a/lib/sql.js b/lib/sql.js
index b51abf6..ea7b77b 100644
--- a/lib/sql.js
+++ b/lib/sql.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var once = require('once');
@@ -14,6 +14,20 @@ var pgError = require('./pg').pgError;
 var control = require('./control');
 
 
+///--- Globals
+
+var ARGS_SCHEMA = [
+    { name: 'statement', type: 'string' },
+    { name: 'values', type: 'array' },
+    { name: 'options', type: 'options' }
+];
+
+var PIPELINE = [
+    control.getPGHandleAndTransaction,
+    execSql
+];
+
+
 ///--- Handlers
 
 function execSql(req, cb) {
@@ -28,7 +42,7 @@ function execSql(req, cb) {
     var q = pg.query(req.stmt, req.values);
 
     q.on('row', function (r) {
-        req.res.write(r);
+        req.rpc.write(r);
     });
 
     q.once('error', cb);
@@ -38,10 +52,18 @@ function execSql(req, cb) {
 
 function sql(options) {
     control.assertOptions(options);
-    var route = 'sql';
 
-    function _sql(stmt, values, opts, res) {
-        var req = control.buildReq(opts, res, options);
+    function _sql(rpc) {
+        var argv = rpc.argv();
+        if (control.invalidArgs(rpc, argv, ARGS_SCHEMA)) {
+            return;
+        }
+
+        var stmt = argv[0];
+        var values = argv[1];
+        var opts = argv[2];
+
+        var req = control.buildReq(opts, rpc, options);
         req.stmt = stmt;
         req.values = values;
 
@@ -49,15 +71,11 @@ function sql(options) {
             stmt: stmt,
             values: values,
             opts: opts
-        }, '%s: entered', route);
+        }, 'sql: entered');
 
         control.handlerPipeline({
-            route: route,
             req: req,
-            funcs: [
-                control.getPGHandle(route, options.manatee, {}, true),
-                execSql
-            ]
+            funcs: PIPELINE
         });
     }
 
diff --git a/lib/standalone.js b/lib/standalone.js
index aa0ef47..0e9e080 100644
--- a/lib/standalone.js
+++ b/lib/standalone.js
@@ -1,4 +1,12 @@
-// Copyright (c) 2013 Joyent, Inc.  All rights reserved.
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2017, Joyent, Inc.
+ */
 
 var EventEmitter = require('events').EventEmitter;
 var util = require('util');
@@ -101,12 +109,7 @@ function Standalone(options) {
 util.inherits(Standalone, EventEmitter);
 
 
-Standalone.prototype.pg = function pg(options, callback) {
-    if (typeof (options) === 'function') {
-        callback = options;
-        options = {};
-    }
-    assert.object(options, 'options');
+Standalone.prototype.pg = function pg(callback) {
     assert.func(callback, 'callback');
 
     // XXX MANTA-734 - take out db selection logic for now and solely
@@ -120,8 +123,7 @@ Standalone.prototype.pg = function pg(options, callback) {
     }
 
     log.debug({
-        db: db,
-        options: options
+        db: db
     }, 'pg: entered');
     db.checkout(function pgCallback(err, client) {
         if (err) {
@@ -138,15 +140,10 @@ Standalone.prototype.pg = function pg(options, callback) {
 };
 
 
-Standalone.prototype.start = function start(opts, cb) {
-    if (typeof (opts) === 'function') {
-        cb = opts;
-        opts = {};
-    }
-
+Standalone.prototype.start = function start(cb) {
     var log = this.log;
 
-    this.pg(opts, function (pool_err, client) {
+    this.pg(function (pool_err, client) {
         if (pool_err) {
             cb(pool_err);
             return;
diff --git a/main.js b/main.js
index 172ce38..c9a0314 100644
--- a/main.js
+++ b/main.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var fs = require('fs');
@@ -17,7 +17,6 @@ var bunyan = require('bunyan');
 var clone = require('clone');
 var extend = require('xtend');
 var getopt = require('posix-getopt');
-var panic = require('panic');
 var VError = require('verror').VError;
 
 
@@ -29,6 +28,7 @@ var app = require('./lib');
 
 var DEFAULTS = {
     file: process.cwd() + '/etc/config.json',
+    kangPort: 3020,
     port: 2020,
     bindip: '0.0.0.0'
 };
@@ -93,7 +93,7 @@ function setupLogger(config) {
 function parseOptions() {
     var option;
     var opts = {};
-    var parser = new getopt.BasicParser('cvf:p:', process.argv);
+    var parser = new getopt.BasicParser('cvf:p:k:', process.argv);
 
     while ((option = parser.getopt()) !== undefined) {
         switch (option.option) {
@@ -115,6 +115,14 @@ function parseOptions() {
             }
             break;
 
+        case 'k':
+            opts.kangPort = parseInt(option.optarg, 10);
+            if (isNaN(opts.kangPort)) {
+                LOG.fatal({ port: option.optarg }, 'Invalid port');
+                throw new Error('Invalid port: ' + option.optarg);
+            }
+            break;
+
         case 'v':
             // Allows us to set -vvv -> this little hackery
             // just ensures that we're never < TRACE
@@ -191,9 +199,4 @@ function run(options) {
             process.exit(0);
         });
     }
-
-    panic.enablePanicOnCrash({
-        'skipDump': true,
-        'abortOnPanic': true
-    });
 })();
diff --git a/package.json b/package.json
index 79362ef..1fbdfc9 100644
--- a/package.json
+++ b/package.json
@@ -6,15 +6,17 @@
     "private": true,
     "main": "lib/index.js",
     "dependencies": {
-        "assert-plus": "0.1.5",
+        "ajv": "4.11.4",
+        "assert-plus": "1.0.0",
         "bunyan": "0.22.1",
         "bunyan-syslog": "0.2.2",
         "clone": "0.1.11",
         "crc": "0.2.1",
         "dtrace-provider": "0.2.8",
         "deep-equal": "0.0.0",
-        "fast": "0.3.6",
+        "fast": "2.2.3",
         "ip6addr": "0.1.1",
+        "kang": "1.2.0",
         "moray-filter": "1.0.0",
         "libuuid": "0.1.3",
         "lru-cache": "2.5.0",
@@ -23,7 +25,6 @@
         "moray": "3.1.1",
         "microtime": "0.5.1",
         "once": "1.3.0",
-        "panic": "0.2.1",
         "pg": "2.11.0",
         "pg-parse-float": "0.0.1",
         "pooling": "0.4.5",
diff --git a/sapi_manifests/moray/template b/sapi_manifests/moray/template
index 9c733ca..6cf4fbb 100644
--- a/sapi_manifests/moray/template
+++ b/sapi_manifests/moray/template
@@ -7,7 +7,6 @@
     }
   },
   "numWorkers": 0,
-  "audit": false,
   "manatee": {
     "manatee": {
       "path": "/manatee/{{SERVICE_NAME}}",
diff --git a/sdc/sapi_manifests/moray/template b/sdc/sapi_manifests/moray/template
index 07ea3a8..a21f367 100644
--- a/sdc/sapi_manifests/moray/template
+++ b/sdc/sapi_manifests/moray/template
@@ -7,7 +7,6 @@
       }
   },
   "numWorkers": 0,
-  "audit": false,
   "manatee": {
     "manatee":{
       "path": "/manatee/{{manatee_shard}}",
diff --git a/smf/manifests/moray.xml.in b/smf/manifests/moray.xml.in
index 6e0d9cf..59b254c 100644
--- a/smf/manifests/moray.xml.in
+++ b/smf/manifests/moray.xml.in
@@ -7,7 +7,7 @@
 -->
 
 <!--
-    Copyright (c) 2014, Joyent, Inc.
+    Copyright (c) 2017, Joyent, Inc.
 -->
 
 <service_bundle type="manifest" name="smartdc-moray">
@@ -44,7 +44,7 @@
         <exec_method
             type="method"
             name="start"
-            exec="node main.js -f etc/config.json -p %{moray/port} &amp;"
+            exec="node --abort-on-uncaught-exception main.js -f etc/config.json -p %{moray/port} -k %{moray/kang} &amp;"
             timeout_seconds="30">
             <method_context working_directory="/opt/smartdc/moray">
                 <method_credential user="nobody"
@@ -70,6 +70,7 @@
         <instance name="@@MORAY_INSTANCE_NAME@@" enabled="true">
           <property_group name="moray" type="application">
             <propval name="port" type="astring" value="@@MORAY_PORT@@" />
+            <propval name="kang" type="astring" value="@@KANG_PORT@@" />
           </property_group>
         </instance>
 
